### ISSUE TITLE
X support

### ISSUE THREAD

**Issue by tpruzina** [FeralInteractive/gamemode#7]
I think gamemode needs some X awareness before we can implement nice features such as:

1. [DVC toggling](https://github.com/NVIDIA/nvidia-settings/blob/master/samples/nv-control-dvc.c) (or equivalent for Intel/AMD - "saturation" knobs)
2. GPU overclocking (see link above for nvidia)
3. Implicit game state tracking (for stuff like disabing game-mode out of focus)
    - Daemon should have awareness of X client (WM_CLASS)
    - It needs listen to X Event loop

When I first seen this, I was kinda excited, hoping that sexy Feral Launcher stuff (display selection) would make it into daemon, but this isn't the case (at least not at the moment).

I could probably fork this and write some PoC nvidia implementation of these over the weekend, but I think I would have to touch whole bunch of core code to make this happen and it would likely get rejected (I don't think including nvidia-settings as gitsubmodule would be a problem, but I haven't looked into licensing issue yet).

**Comment by ikeydoherty** [comment]
X11 awareness sounds like more of an issue for a compositor tbh - and would make Wayland a no-go. I know, I said the Wayland word, laugh away. :P Also don't forget that X11 is incredibly broken when it comes to what truly constitutes a client window (look at gdk 2 + 3) which is why libraries like libwnck exists.

**Comment by tpruzina** [comment]
@ikeydoherty It's crucial for supporting OC/DVC on nvidia as they only support those via their Xorg binary control module to which nvidia API connects to. As for right now, just about every game still depends on X (few SDL2 games don't). This said, If XOpenDisplay() fails (e.g. X is not running), we can just ignore features that aren't available. This wouldn't make it necessarily a run dependency aways, we can always load xlib dynamically via dlopen.

As far as wayland goes, even on wayland you might wanna poll for client state (window focus).

**Comment by ikeydoherty** [comment]
Yeah Wayland won't allow that, hence my remarks on the compositor. "Security" :roll_eyes: Well that would sorta lend the argument to `dlopen()` of backend modules when the codebase starts being multifunctional - and chain the modules for bring "up" and "down" gamemode. The linked code looks pretty simple anyway tbh.

**Comment by tpruzina** [comment]
It's just an sample use case of nvidias api for setting DVC/saturation, what it doesn't do is select correct monitor or anything too fancy.
